dm/docs/RFCS/20211130_enhanced_pre_checker.md

```diff
@@ -0,0 +1,86 @@
+# Enhanced pre-check Design
+
+## Background
+
+Before the DM’s task starts, we should check some items to avoid start-task failures. You can see the details in our [document](https://docs.pingcap.com/tidb-data-migration/stable/precheck#disable-checking-items). In order to allow some users to use DM normally under certain circumstances, we can also set ignore-check-items in task config to ignore some items that you don’t want to check. Now, we find some shortcomings in regard to check-items.
+
+### Bad user habits
+
+We allow users to ignore all check-items, in which case the user's privilege is too large to perform unexpected operations. 
+
+### Too much time overhead
+
+If we have a large number of tables in source, we will take too much time in checking table schema, sharding table consistency and sharding table auto increment ID.
+
+### Inadequate check
+
+* Now we check auto increment ID by `column-mapping` which is deprecated. If we don't set the `column-mapping` and don't ignore the **auto_increment_ID**, the pre-check will report an error. PS: **auto_increment_ID** is only checked when **schema_of_shard_tables** is true.
+* Dump privilege only checks RELOAD and SELECT. However, Dumpling supports different [consistency configurations](https://docs.pingcap.com/tidb/stable/dumpling-overview#adjust-dumplings-data-consistency-options), which need more privileges.
+* If online-ddl is set by true and a DDL is in online-ddl stage, DM will have a problem in all mode. Specifically, when ghost table has been created but not been renamed, DM will report an error when the ghost table is renamed during the incremental phase. You can learn more about online-ddl [here](https://docs.pingcap.com/tidb-data-migration/stable/feature-online-ddl).
+* For schema_of_shard_tables, whether it's a pessimistic task or an optimistic task, we all simply check consistency of schema by comparing all sharding tables’ structures. For optimistic mode, we can do better.
+* Version checker checks that MySQLVersion >= 5.6.0 and MariadbVersion >= 10.1.2 before. However, we find more and more incompatibility problems in MySQLVersion >= 8.0.0 and Mariadb. In view of supporting MySQL 8.0 and Mariadb is experimental yet, checker will report a warning for them.
+
+## Proposal
+
+### Speed ​​up check
+
+1. Support concurrent check
+    - table_schema
+    - schema_of_shard_tables
+    - auto_increment_ID
+2. We can adjust the concurrency by table numbers.
```

Comment:
also can we batch the request by multi-statement or something?

Reply:
just run mutil-thread to execute `show create table`. Can't batch the request. The [draft](https://github.com/pingcap/tiflow/pull/3975)

Reply:
can we use INFORMATION_SCHEMA to get the table structure? @lichunzhu PTAL.

when the number of table is large and we have a high latency network

Reply:
This is the test result from https://github.com/pingcap/tiflow/pull/3975/files#diff-d6ce4d888f567f277d6f1304165d1b3cca440f6563ecb742d7c3a80f7d411b8f

---------------
## 4 threads 10000 tables
|test mechine|added delay|spend time|
| ------------|-----------|---------|
|local|0 ms|3.9 s|
|125|0 ms|8.5 s|
|125|10 ms|31.7 s|
|106|0 ms|11.2 s|
|106|10 ms|29.6 s|
|106|20 ms|54.5 s|
|106|50 ms|128.5 s|
|106|100 ms|253 s|
----------------

I intend to keep the delay under a minute(@sunzhaoyang 's thought).

In other words, if number of table is greater than 5000 and less than 10000, we use 16 threads. If number of table is greater than 10000 and less than 20000, we use 32 threads. The spend time is almost under a minute.

If number is greater than 20000, maybe user can afford the spend time(Will grow linear growth). Or we can use more threads? 64, 128...

Reply:
when we incr check thread, do we also adjust mysql_max_connections to 16/32 ?

Reply:
beacause there is limited underlying  connection count behiend `sql.DB` more thread count than this number(underlying  connection count) can not speed up check process

Reply:
Even if we adjust mysql_max_connection to 16/32, we seem can't guarantee no other connection with this MySQL.

Reply:
yes, so my point is maybe we should just use `mysql.MaxConnectionCount` as concurrent count when need check table num lagre than `mysql.MaxConnectionCount`